ports/rp2: Change MICROPY_PY_LWIP setting value and  netinfo structure for dhcp.

irinakim12 opened the pull request:
Sorry, It occur the problem during forked repository merge operation.
So, I open  re-pull request.

# Issue

1. Modified the register address area each chip in wiznet5k_regs function

-  When MICROPY_PY_LWIP is set as '0' , it can't read the register value of CREG in wiznet5k_regs function.

2. Change the location of netinfo structure in the wiznet5k_init function
3. Delete the define of MICROPY_PY_LWIP  in the mpconfigboard.cmake of W5100S_EVB_PICO and W5500_EVB_PICO.

- When cmake compile, it can't change the value of MICROPY_PY_LWIP

# To do

- [ ] When MICROPY_PY_LWIP is set as '1' , it can read the register value of CREG in wiznet5k_regs function.

- [ ] When MICROPY_PY_LWIP is set as '1' and DHCP use ,the operation of assigning IP is very slow.

irinakim12 committed:
ports/rp2: Change MICROPY_PY_LWIP setting value and netinfo structure.

Delete the define of MICROPY_PY_LWIP in the mpconfigboard.cmake
of W5100S_EVB_PICO and W5500_EVB_PICO.
When cmake compile, it can't change the value of MICROPY_PY_LWIP
Modified the register address area each chip in wiznet5k_regs function.
When MICROPY_PY_LWIP is set as '0' , it can't read the register value
of CREG in wiznet5k_regs function.
Change the location of netinfo structure in the wiznet5k_init function.

irinakim12 committed:
ports/rp2: Change the register value for each chip.

Each chip has a different range of registers.

Signed-off-by: Eunkyoung Kim <irinakim1225@gmail.com>

jimmo commented:
You do not need to close and re-open the pull request. Just force-push to the remote branch that the PR is made from. You can do this from any local branch, even if it's different to the one that the remote branch was initially created from.

I'll re-add the comments from the original PR.

jimmo commented on extmod/network_wiznet5k.c:
```diff
@@ -388,6 +379,17 @@ STATIC void wiznet5k_init(void) {
+    uint8_t sn_size[16] = {2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2};         // 2k buffer for each socket
+    ctlwizchip(CW_INIT_WIZCHIP, sn_size);
+
+    ctlnetwork(CN_SET_NETINFO, (void *)&wiznet5k_obj.netinfo);
```
This change makes sense (calling ctlnetwork after actually initialising netinfo), but maybe only the ctlnetwork line needs to move?

irinakim12 replied:
Right,
In conclusion, after initializing netinfo, call the ctlnetwork function and set the network-related register to the initialized value. Before I change the code , I tried to test it  and I saw it that regs() function call through Thonny and Wiz CREG value hasn't initialized value.

jimmo replied:
Looking into this some more...

The default behavior just about everywhere else in MicroPython (including WizNet-with-LWIP) is to not have a default address, and enable DHCP by default. So while I agree that this code is wrong as it is, I think we should just remove it altogether.

jimmo commented on extmod/network_wiznet5k.c:
```diff
@@ -896,11 +905,11 @@ STATIC mp_obj_t wiznet5k_ifconfig(size_t n_args, const mp_obj_t *args) {
-        netutils_parse_ipv4_addr(items[2], netinfo.gw, NETUTILS_BIG);
-        netutils_parse_ipv4_addr(items[3], netinfo.dns, NETUTILS_BIG);
-        ctlnetwork(CN_SET_NETINFO, &netinfo);
+        netutils_parse_ipv4_addr(items[0], self->netinfo.ip, NETUTILS_BIG);
```
Can you explain this change. I think it makes sense, but it would be good to clarify -- is self->netinfo only for DHCP or is it correct to also put the static address into it? What goes wrong if you don't make this change?

irinakim12 replied:
The existing code doesn't use the IP assigned but it use the IP assigned at the initial.
It is possible to ping test, but when regs() function call, it reads the previous value.

jimmo replied:
I don't understand how this change will affect the regs() function (which reads directly from the wiznet chip).

jimmo replied:
Just ran a test here:

1. Device has default address. Cannot ping.
2. Call nic.ifconfig() with correct static address. Device immediately responds to ping.
3. Call nic.regs(), correctly returns new details.

```
>>> nic.ifconfig(('192.168.1.138', '255.255.255.0', '192.168.1.1', '192.168.1.1'))
>>> nic.regs()
Wiz CREG:
  0000: 00 c0 a8 01 01 ff ff ff 00 02 99 8f 74 61 9a c0
  0010: a8 01 8a 00 00 00 00 00 00 07 d0 08 28 00 00 00
  0020: 00 00 00 00 00 00 ff ff 00 00 00 00 00 00 bf 00
  0030: 00 00 00 00 00 00 00 78 25 04 00 00 00 00 00 00
  0040: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
```

So still not sure why this change is necessary?

irinakim12 replied:
This issue occur the DHCP mode, not Static mode

jimmo commented on ports/rp2/CMakeLists.txt:
```diff
@@ -199,6 +199,10 @@ if (MICROPY_PY_LWIP)
     )
+else()
+	target_compile_definitions(${MICROPY_TARGET} PRIVATE
+        MICROPY_PY_LWIP=0
```
This shouldn't be necessary. The default (unset) is fine.

irinakim12 replied:
Okey, I delete the else statement.

jimmo commented on ports/rp2/Makefile:
```diff
@@ -6,9 +6,14 @@ BOARD ?= PICO
  
 BUILD ?= build-$(BOARD)
 
+LWIP ?= 0
```
I don't think it makes sense to put this into Makefile. If a board wants to configure LWIP/WizNet then that should be done in mpconfigboard.cmake.

This is an unfortunate detail of the way the CMake system works, but I don't think adding each individual option to the Makefile is the right way to solve it.

irinakim12 replied:
The W5100S or W5500 can run in LWIP or non-LWIP mode. 
But WIZnet's Chip is Hardwared TCP/IP so, the non-LWIP mode is basic operation.
If you were, where would you configure the code to be selected(LWIP or non-LWIP)?

irinakim12 replied:
I have never received a PR. So, If you tell me the location of BOARD_VARIANT or how to apply it, I will change it.

jimmo commented on ports/rp2/boards/W5100S_EVB_PICO/mpconfigboard.cmake:
```diff
@@ -1,4 +1,3 @@
 # cmake file for Wiznet W5100S-EVB-Pico.
 set(PICO_BOARD wiznet_w5100s_evb_pico)
 set(MICROPY_PY_NETWORK_WIZNET5K W5100S)
-set(MICROPY_PY_LWIP 1)
```
These should stay (unless there's some reason we want to stop using LWIP on these boards?) @andrewleech ?

irinakim12 replied:
First, the chip of WIZNET must run its own hardware TCP/IP by default.
Working with LWIP should be an option.

Therefore, if a value of 1 is given to MICROPY_PY_LWIP in the mpconfigboard.cmake , even if the LWIP(=MICROPY_PY_LWIP) value is set to '0' as an option when running make, the MICROPY_PY_LWIP value is eventually changed to '1' by the value of mpconfigboard.cmake.

For reference, when operating in LWIP, Ethernet operation speed is slow.

jimmo replied:
> First, the chip of WIZNET must run its own hardware TCP/IP by default.

Why? Is it just the speed difference? How much difference does it make?

From MicroPython's perspective, for consistency, it's better if we use the same stack on all boards and ports.

jimmo replied:
I just did some performance tests.

With the SPI baudrate at 20M:
  LWIP: 5.5Mbps.
  Hardware TCP: 10.5Mbps

At 10M:
  LWIP: 3.7Mbps
  Hardware TCP: 6.6Mbps

So about twice as fast. @irinakim12 why is this? Is it a hardware limitation or a driver issue? I'm curious why the SPI baudrate affects it like this, is there a (very!) significant per-frame overhead in the raw mode?

@andrewleech Do you know more about this?

andrewleech replied:
When I was testing the initial integration I did find the hardware TCP stack to be faster as found, but somewhat more limited - lwip can have any number of sockets (only limited by ram) and a "full" range of other network features. The hardware TCP bindings were missing a few features, though I think I added a couple of them to satisfy unit tests... can't remember for sure now.

I never really investigated the speed difference of the two stacks, though some of it was due / affected by the background polling of lwip. I'd attempted to improve on that with the interrupts, though didn't thoroughly profile the remaining speed differences.

SPI clock affecting speed appears to be expected though: https://docs.wiznet.io/Product/iEthernet/W5500/Application/spi-performance#data-throughput-according-to-spi-clock-speed

irinakim12 replied:
@jimmo  I also tried to modify the code because the speed slowed down when operating with LWIP, but I don't know where the lock takes place.  Since it is a hardware device, it is unavoidably limited, but I understand that 8 channels are available, and the current code is coded to use only 1 channel.

irinakim12 committed:
ports/rp2: Delete the else statement.

Delete the else statement about MICROPY_PY_LWIP.
Because MICROPY_PY_LWIP is set '0' at the initial,
it needs no else statement.

Signed-off-by: Eunkyoung Kim <irinakim1225@gmail.com>

irinakim12 reviewed (commented):
I've finished the my review..

irinakim12 commented:
@dpgeorge @jimmo   Hi~ I look forward to hearing your answers and confirmation.
